src/replication.c

```diff
@@ -2193,8 +2217,10 @@ void readSyncBulkPayload(connection *conn) {
             bg_unlink(server.rdb_filename);
         }
 
+        /* Reclaim the cache backed by rdb */
+        bioCreateCloseJob(server.repl_transfer_fd, 0, 1);
```

Comment:
Maybe we'd better move the cache reclaim to `rdbLoad` just like what `rdbSave` does, but `rdbLoad` operate with `FILE` while `bioCreateCloseJob` operate with fd, I don't like open and close a file twice, maybe we could refactor `FILE` to fd and maintain the file buffer ourself in the future.

Reply:
you could use `fileno` to get the fd of that FILE, but i'm not certain we want to move it to rdbLoad.

readSyncBulkPayload is the one that writes that file, and also calls rdbLoad to read it, so at least the writing concern belongs here (also it's the one that orchestrates the load).

maybe the question should be if we want to reclaim the cache after loading an rdb file on startup.
remember we did RDBFLAGS_KEEP_CACHE in finishShutdown.
since unlike other cases, in both of these cases we don't incrementally reclaim, and instead we can allow the entire file to get cached, it could be that none of this really helps reducing the damage.
also, if we reclaim after loading at startup, we should make sure to do it only on successful load, since it could be that loading fails due to some bad configuration, and it's gonna be re-attempted.

IMHO we can keep it as it is now.

Reply:
> you could use fileno to get the fd of that FILE

In the bio we would fadvise and close the fd, outside the bio(i.e. in rdbLoad) we close the FILE, as there may have race condition in two different thread, so we'd better do all the file operations in bio, but I guess close fd is not equal to close a FILE(at least the memory refered by FILE pointer is not released), so passing the fileno of FILE to bioCreateCloseJob is not enough.

> since unlike other cases, in both of these cases we don't incrementally reclaim, and instead we can allow the entire file to get cached, it could be that none of this really helps reducing the damage.

Specifically in our use case, a high-capacity host carrying tens or hundreds of redis instances, as page cache in each redis instance grows, it assembles substantial page cache in the host(usually more than 50GB), and finally triggers direct reclaim. As the number of pages to be scanned and reclaimed is quite a lot, the stall time is not ignorable. To solve this specific case finally reclaiming the cache after load is ok. In low-capacity machine the OS needn't do many works to relcaim cache as the total pages is less, so this won't be a issue.

Reply:
you're right.
so you wanna add a reclaim on startup when an RDB was successfully loaded?

Reply:
i don't think an extra open (at startup) is an issue.